### PR TITLE
PIM-9715: Prevent the deletion of an attribute used as a label by a family

### PR DESCRIPTION
**Pull request by anaisbaunelemaire**
**Description (for Contributor and Core Developer)**

When an attribute is used as a label by a family, you could delete it.

To prevent this, I have created 2 methods that check if the attribute that we want to remove is used as a label in a family and, if so, an error message is displayed for the user: _"Attributes used as labels in a family cannot be removed."_

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Family/CheckAttributeOnDeletionSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        $attributes = array_filter(
+            $attributes,
+            fn ($attr): bool => $attr instanceof AttributeInterface
+        );
+        if ([] === $attributes) {
+            return;
+        }
+
+        foreach ($attributes as $attribute) {
+            if ($attribute instanceof FamilyAttributeAsLabel) {
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Family/CheckAttributeOnDeletionSubscriber.php:52]: FamilyAttributeAsLabel is a validation constraint, this test will always return false
here you should make a database query to retrieve families which have this attribute as label

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Family/CheckAttributeOnDeletionSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+     * Check if the attribute is used as label for a family
+     *
+     * @param RemoveEvent $event
+     * @throws AttributeAsLabelException
+     */
+    public function preRemove(RemoveEvent $event)
+    {
+        $attributes = $event->getSubject();
+
+        if (!is_array($attributes)) {
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Family/CheckAttributeOnDeletionSubscriber.php:40]: what about unitary removal?

File: src/Akeneo/Pim/Enrichment/Component/Exception/AttributeAsLabelException.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace Akeneo\Pim\Enrichment\Component\Exception;
+
+/**
+ * Exception raises when try to remove an attribute used as label for any family
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Exception/AttributeAsLabelException.php:6]:
```suggestion
 * Exception raised when trying to remove an attribute used as label for any family
```

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Family/CheckAttributeOnDeletionSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Akeneo\Pim\Enrichment\Bundle\EventSubscriber\Family;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Family/CheckAttributeOnDeletionSubscriber.php:5]: should be moved to `Akeneo\Pim\Structure\Bundle\EventSubscriber` IMO

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Family/CheckAttributeOnDeletionSubscriber.php (hunk 1 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Family/CheckAttributeOnDeletionSubscriber.php:52]: e.g: 
```sql
SELECT EXISTS(
    SELECT * FROM pim_catalog_family
    WHERE label_attribute_id = :attributeId
) as attribute_is_used;
```

**Review by olivierpontier — APPROVED** [review]

**Review by ValentinMumble — APPROVED** [review]

**Pull request merged**